service/controller/resource/ipam/create.go

```diff
@@ -69,8 +71,21 @@ func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
 	// allocation, if the NetworkPool CR is given.
 	var networkRange net.IPNet
 	{
+		var clusterID string
+
+		// We need to check if the request comes from a Cluster or a MachineDeployment in order to fetch the name of the Cluster accordingly
+		clusterID = m.GetName()
+		if reflect.TypeOf(r.checker) == reflect.TypeOf(MachineDeploymentChecker{}) {
+			var md v1alpha2.AWSMachineDeployment
+			err = r.k8sClient.CtrlClient().Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, &md)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			clusterID = md.Labels[key.TagCluster]
+		}
+
 		var cr v1alpha2.AWSCluster
-		err = r.k8sClient.CtrlClient().Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, &cr)
+		err = r.k8sClient.CtrlClient().Get(ctx, types.NamespacedName{Name: clusterID, Namespace: m.GetNamespace()}, &cr)
```

Comment:
I do not follow why the reflect magic above is necessary. Can't we simply do `key.Cluster(m)`? The meta accessor provides `GetLabels` and both the `Cluster` CR and the `MachineDeployment` CR have the cluster label. So if we only need the cluster ID the key function should do it. No reflection needed.

Reply:
Oh man, here I was trying to do it the complicated way! I will test this now :)

Reply:
7/6 for creative solution.